clients/pubsub-common/src/main/java/org/eclipse/hono/client/pubsub/PubSubMessageHelper.java

```diff
@@ -70,12 +74,76 @@ private static GoogleCredentials getCredentials() {
     /**
      * Gets the topic name with the given prefix.
      *
-     * @param topic The endpoint of the topic (e.g. event)
-     * @param prefix The prefix of the Pub/Sub topic, it's either the tenant ID or the adapter instance ID
+     * @param topic The endpoint of the topic (e.g. event).
+     * @param prefix The prefix of the Pub/Sub topic, it's either the tenant ID or the adapter instance ID.
      * @return The topic containing the prefix identifier and the endpoint.
      */
     public static String getTopicName(final String topic, final String prefix) {
-        return String.format("%s.%s", prefix, topic);
+        return getTopicName(topic, prefix, Collections.emptyList());
+    }
+
+    /**
+     * Gets the topic name with the given prefix and subtopics.
+     *
+     * @param topic The endpoint of the topic (e.g. event).
+     * @param prefix The prefix of the Pub/Sub topic, it's either the tenant ID or the adapter instance ID.
+     * @param subtopics A list with the subtopics.
+     * @return The topic containing the prefix identifier, the endpoint and all the subtopics.
+     */
+    public static String getTopicName(final String topic, final String prefix, final List<String> subtopics) {
+        final StringBuilder topicBuilder = new StringBuilder();
+        topicBuilder.append(prefix).append(".").append(topic);
+        for (String subtopic : subtopics) {
+            topicBuilder.append(".").append(subtopic);
+        }
+        return topicBuilder.toString();
+    }
+
+    /**
+     * Gets the subtopics from the orig_address attribute of the message.
+     *
+     * @param origAddress The orig_address attribute.
+     * @return A list containing all the subtopics in hierarchical order or an empty list if the topic has no subtopics.
+     */
+    public static List<String> getSubtopics(final String origAddress) {
```

Comment:
FMPOV this should return an *ImmutableList<String>* and adapt the JavaDoc accordingly.

Reply:
I would rather not make this change since Collections.emptyList() returns a List<T> rather than an ImmutableList<T>. Otherwise we would have to convert both the empty list aswell as origAddressSplit into ImmutableLists.

Reply:
I was not so much referring to returning Guava's ImmutableList but to return a List wrapped in *Collections.unmodifiableList* and make it clear in the JavaDoc that the list being returned cannot be modified.

Reply:
ok got it. It should be fixed now.